# Building a Spam Filter with Naive Bayes
## Introduction
We're going to build a spam filter for SMS messages using the multinomial Naive Bayes algorithm.

To classify messages as spam or non-spam, the computer:
1. Learns how humans classify messages.
2. Uses that human knowledge to estimate probabilities for new messages — probabilities for spam and non-spam.
3. Classifies a new message based on these probability values — if the probability for spam is greater, then it classifies the message as spam. Otherwise, it classifies it as non-spam (if the two probability values are equal, then we may need a human to classify the message).

To train the algorithm, we'll use a dataset of 5,572 SMS messages that are already classified by humans. The dataset can be downloaded from the the [UCI Machine Learning Repository](https://archive.ics.uci.edu/ml/datasets/sms+spam+collection).

In [1]:
import pandas as pd
sms_spam = pd.read_csv('SMSSpamCollection', sep = '\t', header = None, names = ['Label', 'SMS'])
sms_spam

Label                                                SMS
0      ham  Go until jurong point, crazy.. Available only ...
1      ham                      Ok lar... Joking wif u oni...
2     spam  Free entry in 2 a wkly comp to win FA Cup fina...
3      ham  U dun say so early hor... U c already then say...
4      ham  Nah I don't think he goes to usf, he lives aro...
...    ...                                                ...
5567  spam  This is the 2nd time we have tried 2 contact u...
5568   ham               Will ü b going to esplanade fr home?
5569   ham  Pity, * was in mood for that. So...any other s...
5570   ham  The guy did some bitching but I acted like i'd...
5571   ham                         Rofl. Its true to its name

[5572 rows x 2 columns]

In [2]:
sms_spam['Label'].value_counts(normalize = True) * 100

ham     86.593683
spam    13.406317
Name: Label, dtype: float64

We see that about 87% of the messages are ham, and the remaining 13% are spam. This sample looks representative, since in practice most messages that people receive are ham.

## Training and Test Set
We're now going to split our dataset into a training and a test set, where the training set accounts for 80% of the data, and the test set for the remaining 20%.

In [3]:
randomized_data = sms_spam.sample(frac = 1)
train_set_length = round(len(randomized_data) * 0.8)
test_set_length = len(randomized_data) - train_set_length
train_set = randomized_data.iloc[:train_set_length].reset_index(drop = True)
test_set = randomized_data.iloc[train_set_length:].reset_index(drop = True)

In [4]:
train_set['Label'].value_counts(normalize = True) * 100

ham     86.765366
spam    13.234634
Name: Label, dtype: float64

In [5]:
test_set['Label'].value_counts(normalize = True) * 100

ham     85.906643
spam    14.093357
Name: Label, dtype: float64

## Data Cleaning
To calculate all the probabilities required by the algorithm, we'll first need to perform a bit of data cleaning to bring the data in a format that will allow us to extract easily all the information we need.
### Letter Case and Punctuation
We'll begin with removing all the punctuation and bringing every letter to lower case.

In [6]:
train_set['SMS'] = train_set['SMS'].str.replace('\W', ' ')
train_set['SMS'] = train_set['SMS'].str.lower()
train_set.head()

Label                                                SMS
0   ham                          india have to take lead  
1   ham  your account has been refilled successfully by...
2   ham           otherwise had part time job na tuition  
3   ham       im just wondering what your doing right now 
4   ham  yo carlos  a few friends are already asking me...

### Creating the Vocabulary
Let's now move to creating the vocabulary, which in this context means a list with all the unique words in our training set.

In [7]:
train_set['SMS'] = train_set['SMS'].str.split()
vocabulary = []
for sms in train_set['SMS']:
    for word in sms:
        vocabulary.append(word)
        
vocabulary = list(set(vocabulary))
len(vocabulary)

7822

### The Final Training Set
We're now going to use the vocabulary we just created to make the data transformation we want.

In [8]:
word_counts_per_sms = {unique_word: [0] * len(train_set['SMS']) for unique_word in vocabulary}
for index, sms in enumerate(train_set['SMS']):
    for word in sms:
        word_counts_per_sms[word][index] += 1

In [9]:
word_counts = pd.DataFrame(word_counts_per_sms)
word_counts

grinder  dob  arms  agents  kalisidare  730  mandan  oblivious  q  \
0           0    0     0       0           0    0       0          0  0   
1           0    0     0       0           0    0       0          0  0   
2           0    0     0       0           0    0       0          0  0   
3           0    0     0       0           0    0       0          0  0   
4           0    0     0       0           0    0       0          0  0   
...       ...  ...   ...     ...         ...  ...     ...        ... ..   
4453        0    0     0       0           0    0       0          0  0   
4454        0    0     0       0           0    0       0          0  0   
4455        0    0     0       0           0    0       0          0  0   
4456        0    0     0       0           0    0       0          0  0   
4457        0    0     0       0           0    0       0          0  0   

      shattered  ...  achieve  trivia  chatlines  serena  edison  checked  \
0             0  ...        0       0          0       0       0        0   
1             0  ...        0       0          0       0       0        0   
2             0  ...        0       0          0       0       0        0   
3             0  ...        0       0          0       0       0        0   
4             0  ...        0       0          0       0       0        0   
...         ...  ...      ...     ...        ...     ...     ...      ...   
4453          0  ...        0       0          0       0       0        0   
4454          0  ...        0       0          0       0       0        0   
4455          0  ...        0       0          0       0       0        0   
4456          0  ...        0       0          0       0       0        0   
4457          0  ...        0       0          0       0       0        0   

      brb  trial  phoned  max10mins  
0       0      0       0          0  
1       0      0       0          0  
2       0      0       0          0  
3       0      0       0          0  
4       0      0       0          0  
...   ...    ...     ...        ...  
4453    0      0       0          0  
4454    0      0       0          0  
4455    0      0       0          0  
4456    0      0       0          0  
4457    0      0       0          0  

[4458 rows x 7822 columns]

In [10]:
train_set_clean = pd.concat([train_set, word_counts], axis = 1)
train_set_clean

Label                                                SMS  grinder  dob  \
0      ham                      [india, have, to, take, lead]        0    0   
1      ham  [your, account, has, been, refilled, successfu...        0    0   
2      ham     [otherwise, had, part, time, job, na, tuition]        0    0   
3      ham  [im, just, wondering, what, your, doing, right...        0    0   
4      ham  [yo, carlos, a, few, friends, are, already, as...        0    0   
...    ...                                                ...      ...  ...   
4453   ham  [pls, confirm, the, time, to, collect, the, ch...        0    0   
4454  spam  [xmas, new, years, eve, tickets, are, now, on,...        0    0   
4455   ham                     [like, lt, gt, same, question]        0    0   
4456   ham                 [thats, cool, how, was, your, day]        0    0   
4457   ham  [my, ex, wife, was, not, able, to, have, kids,...        0    0   

      arms  agents  kalisidare  730  mandan  oblivious  ...  achieve  trivia  \
0        0       0           0    0       0          0  ...        0       0   
1        0       0           0    0       0          0  ...        0       0   
2        0       0           0    0       0          0  ...        0       0   
3        0       0           0    0       0          0  ...        0       0   
4        0       0           0    0       0          0  ...        0       0   
...    ...     ...         ...  ...     ...        ...  ...      ...     ...   
4453     0       0           0    0       0          0  ...        0       0   
4454     0       0           0    0       0          0  ...        0       0   
4455     0       0           0    0       0          0  ...        0       0   
4456     0       0           0    0       0          0  ...        0       0   
4457     0       0           0    0       0          0  ...        0       0   

      chatlines  serena  edison  checked  brb  trial  phoned  max10mins  
0             0       0       0        0    0      0       0          0  
1             0       0       0        0    0      0       0          0  
2             0       0       0        0    0      0       0          0  
3             0       0       0        0    0      0       0          0  
4             0       0       0        0    0      0       0          0  
...         ...     ...     ...      ...  ...    ...     ...        ...  
4453          0       0       0        0    0      0       0          0  
4454          0       0       0        0    0      0       0          0  
4455          0       0       0        0    0      0       0          0  
4456          0       0       0        0    0      0       0          0  
4457          0       0       0        0    0      0       0          0  

[4458 rows x 7824 columns]

## Calculating Constants First
We're now done with cleaning the training set, and we can begin creating the spam filter. The Naive Bayes algorithm will need to answer these two probability questions to be able to classify new messages:

$$
P(Spam | w_1,w_2, ..., w_n) \propto P(Spam) \cdot \prod_{i=1}^{n}P(w_i|Spam)
$$$$
P(Ham | w_1,w_2, ..., w_n) \propto P(Ham) \cdot \prod_{i=1}^{n}P(w_i|Ham)
$$
Also, to calculate P(wi|Spam) and P(wi|Ham) inside the formulas above, we'll need to use these equations:

$$
P(w_i|Spam) = \frac{N_{w_i|Spam} + \alpha}{N_{Spam} + \alpha \cdot N_{Vocabulary}}
$$$$
P(w_i|Ham) = \frac{N_{w_i|Ham} + \alpha}{N_{Ham} + \alpha \cdot N_{Vocabulary}}
$$
Some of the terms in the four equations above will have the same value for every new message. We can calculate the value of these terms once and avoid doing the computations again when a new messages comes in. Below, we'll use our training set to calculate:

P(Spam) and P(Ham)
NSpam, NHam, NVocabulary
We'll also use Laplace smoothing and set $\alpha = 1$

In [11]:
spam_messages = train_set_clean[train_set_clean['Label'] == 'spam']
ham_messages = train_set_clean[train_set_clean['Label'] == 'ham']

p_spam = len(spam_messages) / len(train_set_clean)
p_ham = len(ham_messages) / len(train_set_clean)

n_spam = spam_messages['SMS'].apply(len).sum()
n_ham = ham_messages['SMS'].apply(len).sum()

n_vocabulary = len(vocabulary)
alpha = 1

## Calculating Parameters
Now that we have the constant terms calculated above, we can move on with calculating the parameters $P(w_i|Spam)$ and $P(w_i|Ham)$. Each parameter will thus be a conditional probability value associated with each word in the vocabulary.

The parameters are calculated using the formulas:

$$
P(w_i|Spam) = \frac{N_{w_i|Spam} + \alpha}{N_{Spam} + \alpha \cdot N_{Vocabulary}}
$$$$
P(w_i|Ham) = \frac{N_{w_i|Ham} + \alpha}{N_{Ham} + \alpha \cdot N_{Vocabulary}}
$$

In [12]:
parameters_spam = {unique_word: 0 for unique_word in vocabulary}
parameters_ham = {unique_word: 0 for unique_word in vocabulary}

for word in vocabulary:
    n_word_given_spam = spam_messages[word].sum()
    p_word_given_spam = (n_word_given_spam + alpha) / (n_spam + alpha * n_vocabulary)
    parameters_spam[word] = p_word_given_spam
    
    n_word_given_ham = ham_messages[word].sum()
    p_word_given_ham = (n_word_given_ham + alpha) / (n_ham + alpha * n_vocabulary)
    parameters_ham[word] = p_word_given_ham

## Classifying A New Message
Now that we have all our parameters calculated, we can start creating the spam filter. The spam filter can be understood as a function that:
- Takes in as input a new message (w1, w2, ..., wn).
- Calculates P(Spam|w1, w2, ..., wn) and P(Ham|w1, w2, ..., wn).
- Compares the values of P(Spam|w1, w2, ..., wn) and P(Ham|w1, w2, ..., wn), and:
 - If P(Ham|w1, w2, ..., wn) > P(Spam|w1, w2, ..., wn), then the message is classified as ham.
 - If P(Ham|w1, w2, ..., wn) < P(Spam|w1, w2, ..., wn), then the message is classified as spam.
 - If P(Ham|w1, w2, ..., wn) = P(Spam|w1, w2, ..., wn), then the algorithm may request human help.

In [13]:
import re

def classify(message):
    message = re.sub('\W', ' ', message)
    message = message.lower()
    message = message.split()
    
    p_spam_given_message = p_spam
    p_ham_given_message = p_ham
    
    for word in message:
        if word in parameters_spam:
            p_spam_given_message *=  parameters_spam[word]
        if word in parameters_ham:
            p_ham_given_message *= parameters_ham[word]
    
    if p_ham_given_message > p_spam_given_message:
        return 'ham'
    elif p_ham_given_message < p_spam_given_message:
        return 'spam'
    else:
        return 'Equal probabilities, have aa human classify this!'

In [14]:
classify('WINNER!! This is the secret code to unlock the money: C3421.')

'spam'

In [18]:
classify("Im vaccinatedd !")

'ham'

## Measuring the Spam Filter's Accuracy
The two results above look promising, but let's see how well the filter does on our test set, which has 1,114 messages.

We'll start by writing a function that returns classification labels instead of printing them.

In [16]:
test_set['predicted'] = test_set['SMS'].apply(classify)

In [17]:
correct = 0
total = test_set.shape[0]
    
for row in test_set.iterrows():
    row = row[1]
    if row['Label'] == row['predicted']:
        correct += 1
        
print('Correct:', correct)
print('Incorrect:', total - correct)
print('Accuracy:', correct/total)

Correct: 1103
Incorrect: 11
Accuracy: 0.9901256732495511


The accuracy is close to 99%, which is really good. Our spam filter looked at 1,114 messages that it hasn't seen in training, and classified 1,103 correctly.